server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -207,9 +208,20 @@ public void onRemoval(RemovalNotification<Key, BytesReference> notification) {
         // shards as part of request cache.
         Key key = notification.getKey();
         cacheEntityLookup.apply(key.shardId).ifPresent(entity -> entity.onRemoval(notification));
-        cacheCleanupManager.updateCleanupKeyToCountMapOnCacheEviction(
-            new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId)
-        );
+        if (notificationAffectsStaleness(notification)) {
+            CleanupKey cleanupKey = new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId);
```

Comment:
In what cases `cacheEntity` can be null here ? In null case we are not doing any update to the `cleanupKeyToCountMap`

Reply:
You can move the if condition inside the update method that way staleness related logic remains in single place

Reply:
make sense.

Reply:
>In what cases cacheEntity can be null here ? In null case we are not doing any update to the cleanupKeyToCountMap

The only case where cacheEntity can be null here is when the shard has been `deleted`

Reply:
So in that we will leak the values in the `cleanupKeyToCountMap` ?